File: prosessering/src/test/java/no/nav/familie/prosessering/internal/AsyncTaskTest.java
package no.nav.familie.prosessering.internal;

import no.nav.familie.prosessering.AsyncTask;
import no.nav.familie.prosessering.TaskBeskrivelse;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.aop.framework.AopProxyUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@DataJpaTest(excludeAutoConfiguration = {FlywayAutoConfiguration.class})
public class AsyncTaskTest {

    @Autowired
    private List<AsyncTask> tasker;

    @Test
    public void skal_ha_annotasjon() {
        assertThat(tasker.stream().anyMatch(this::harIkkePåkrevdAnnotasjon)).isFalse();
    }

    private boolean harIkkePåkrevdAnnotasjon(AsyncTask it) {
        return !AnnotationUtils.isAnnotationDeclaredLocally(TaskBeskrivelse.class, it.getClass());
    }

    @Test
    public void skal_ha_unike_nav() {
        final var taskTyper = tasker.stream().map(this::finnAnnotasjon).map(TaskBeskrivelse::taskType).collect(Collectors.toList());

        assertThat(taskTyper).isEqualTo(taskTyper.stream().distinct().collect(Collectors.toList()));
    }

    private TaskBeskrivelse finnAnnotasjon(AsyncTask task) {
        final Class<?> aClass = AopProxyUtils.ultimateTargetClass(task);
        return AnnotationUtils.findAnnotation(aClass, TaskBeskrivelse.class);
    }
}

File: prosessering/src/test/java/no/nav/familie/prosessering/internal/AsyncTaskTest.kt
package no.nav.familie.prosessering.internal

import no.nav.familie.prosessering.AsyncTask
import no.nav.familie.prosessering.TaskBeskrivelse
import org.assertj.core.api.Assertions
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.aop.framework.AopProxyUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.core.annotation.AnnotationUtils
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@DataJpaTest(excludeAutoConfiguration = [FlywayAutoConfiguration::class])
class AsyncTaskTest {

    @Autowired
    private lateinit var tasker: List<AsyncTask>

    @Test
    fun `skal ha annotasjon`() {
        Assertions.assertThat(tasker.any {
            harIkkePåkrevdAnnotasjon(it)
        }).isFalse()
    }

    private fun harIkkePåkrevdAnnotasjon(it: AsyncTask): Boolean {
        return !AnnotationUtils.isAnnotationDeclaredLocally(TaskBeskrivelse::class.java,
                                                            it.javaClass)
    }

    @Test
    fun `skal ha unike navn`() {
        val taskTyper = tasker.map { task: AsyncTask -> finnAnnotasjon(task).tasktype }

        Assertions.assertThat(taskTyper)
                .isEqualTo(taskTyper.distinct())
    }

    private fun finnAnnotasjon(task: AsyncTask): TaskBeskrivelse {
        val aClass = AopProxyUtils.ultimateTargetClass(task)
        return AnnotationUtils.findAnnotation(aClass, TaskBeskrivelse::class.java) ?: error("")
    }
}
